File: sherlock_project/langtrace.rb
require 'net/http'
require 'json'

class LangsmithTraceBuilder
  attr_reader :username, :site_data, :proxy, :timeout

  def initialize(username, site_data, proxy: nil, timeout: 60)
    @username = username
    @site_data = site_data
    @proxy = proxy
    @timeout = timeout
  end

  def build_trace
    results_total = {}
    site_data.each do |site_name, site_info|
      url = site_info['url'].gsub('{username}', username)
      response = fetch_url(url)
      results_total[site_name] = analyze_response(response)
    end
    results_total
  end

  private

  def fetch_url(url)
    uri = URI(url)
    http = Net::HTTP.new(uri.host, uri.port, proxy)
    http.read_timeout = timeout
    request = Net::HTTP::Get.new(uri.request_uri)
    http.request(request)
  end

  def analyze_response(response)
    {
      status_code: response.code,
      body: response.body,
      headers: response.to_hash
    }
  end
end

# Example usage:
# site_data = {
#   "example_site" => { "url" => "https://example.com/{username}" }
# }
# trace_builder = LangsmithTraceBuilder.new("testuser", site_data)
# trace = trace_builder.build_trace
# puts trace

File: sherlock_project/langtrace.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"net/url"
	"time"
)

type LangsmithTraceBuilder struct {
	Username string
	SiteData map[string]map[string]string
	Proxy    *url.URL
	Timeout  time.Duration
}

func NewLangsmithTraceBuilder(username string, siteData map[string]map[string]string, proxy *url.URL, timeout time.Duration) *LangsmithTraceBuilder {
	return &LangsmithTraceBuilder{
		Username: username,
		SiteData: siteData,
		Proxy:    proxy,
		Timeout:  timeout,
	}
}

func (ltb *LangsmithTraceBuilder) BuildTrace() (map[string]map[string]interface{}, error) {
	resultsTotal := make(map[string]map[string]interface{})
	for siteName, siteInfo := range ltb.SiteData {
		url := siteInfo["url"]
		url = replaceUsername(url, ltb.Username)
		response, err := ltb.fetchURL(url)
		if err != nil {
			return nil, err
		}
		resultsTotal[siteName] = ltb.analyzeResponse(response)
	}
	return resultsTotal, nil
}

func replaceUsername(url, username string) string {
	return strings.Replace(url, "{username}", username, -1)
}

func (ltb *LangsmithTraceBuilder) fetchURL(urlStr string) (*http.Response, error) {
	client := &http.Client{
		Timeout: ltb.Timeout,
	}
	if ltb.Proxy != nil {
		client.Transport = &http.Transport{
			Proxy: http.ProxyURL(ltb.Proxy),
		}
	}
	req, err := http.NewRequest("GET", urlStr, nil)
	if err != nil {
		return nil, err
	}
	return client.Do(req)
}

func (ltb *LangsmithTraceBuilder) analyzeResponse(response *http.Response) map[string]interface{} {
	defer response.Body.Close()
	body, _ := ioutil.ReadAll(response.Body)
	headers := make(map[string][]string)
	for k, v := range response.Header {
		headers[k] = v
	}
	return map[string]interface{}{
		"status_code": response.StatusCode,
		"body":        string(body),
		"headers":     headers,
	}
}

func main() {
	siteData := map[string]map[string]string{
		"example_site": {"url": "https://example.com/{username}"},
	}
	traceBuilder := NewLangsmithTraceBuilder("testuser", siteData, nil, 60*time.Second)
	trace, err := traceBuilder.BuildTrace()
	if err != nil {
		fmt.Println("Error building trace:", err)
		return
	}
	traceJSON, _ := json.MarshalIndent(trace, "", "  ")
	fmt.Println(string(traceJSON))
}